File: python_problems/TWOBARS.py
from s2mpjlib import *
class  TWOBARS(CUTEst_problem):

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# 
# 
#    Problem:
#    ********
# 
#    Structureal analysis of the simplest two bar scheme.  The structure has
#    the following simple symmetric shape
# 
#                                 *
#                                / \
#                               /   \
#                              /     \
#                            """     """
# 
#    and a force is applied at the top node.  The unknown are the distance
#    of the left and right feet wrt to the projection of the top node and the
#    weight of the bars.
# 
#    Source:
#    an example in a talk by W.K. Zhang and C. Fleury, LLN, 1994.
# 
#    SIF input: Ph. Toint, November 1994
# 
#    classification = "C-COOR2-MN-2-2"
# 
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#   Translated to Python by S2MPJ version 7 II 2026
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    name = 'TWOBARS'

    def __init__(self, *args): 
        import numpy as np
        from scipy.sparse import csr_matrix
        nargin   = len(args)

        #%%%%%%%%%%%%%%%%%%%  PREAMBLE %%%%%%%%%%%%%%%%%%%%
        v_  = {}
        ix_ = {}
        ig_ = {}
        #%%%%%%%%%%%%%%%%%%%  VARIABLES %%%%%%%%%%%%%%%%%%%%
        self.xnames = np.array([])
        self.xscale = np.array([])
        intvars   = np.array([])
        binvars   = np.array([])
        irA          = np.array([],dtype=int)
        icA          = np.array([],dtype=int)
        valA         = np.array([],dtype=float)
        [iv,ix_,_] = s2mpj_ii('X1',ix_)
        self.xnames=arrset(self.xnames,iv,'X1')
        [iv,ix_,_] = s2mpj_ii('X2',ix_)
        self.xnames=arrset(self.xnames,iv,'X2')
        #%%%%%%%%%%%%%%%%%%  DATA GROUPS %%%%%%%%%%%%%%%%%%%
        self.gscale  = np.array([])
        self.grnames = np.array([])
        cnames       = np.array([])
        self.cnames  = np.array([])
        gtype        = np.array([])
        [ig,ig_,_] = s2mpj_ii('OBJ',ig_)
        gtype = arrset(gtype,ig,'<>')
        [ig,ig_,_] = s2mpj_ii('CONS1',ig_)
        gtype = arrset(gtype,ig,'<=')
        cnames = arrset(cnames,ig,'CONS1')
        [ig,ig_,_] = s2mpj_ii('CONS2',ig_)
        gtype = arrset(gtype,ig,'<=')
        cnames = arrset(cnames,ig,'CONS2')
        #%%%%%%%%%%%%%% GLOBAL DIMENSIONS %%%%%%%%%%%%%%%%%
        self.n   = len(ix_)
        ngrp   = len(ig_)
        legrps = np.where(gtype=='<=')[0]
        eqgrps = np.where(gtype=='==')[0]
        gegrps = np.where(gtype=='>=')[0]
        self.nle = len(legrps)
        self.neq = len(eqgrps)
        self.nge = len(gegrps)
        self.m   = self.nle+self.neq+self.nge
        self.congrps = np.concatenate((legrps,eqgrps,gegrps))
        self.cnames = cnames[self.congrps]
        self.nob = ngrp-self.m
        self.objgrps = np.where(gtype=='<>')[0]
        #%%%%%%%%%%%%%%%%%% CONSTANTS %%%%%%%%%%%%%%%%%%%%%
        self.gconst = np.zeros((ngrp,1))
        self.gconst = arrset(self.gconst,ig_['CONS1'],float(1.0))
        self.gconst = arrset(self.gconst,ig_['CONS2'],float(1.0))
        #%%%%%%%%%%%%%%%%%%%%  BOUNDS %%%%%%%%%%%%%%%%%%%%%
        self.xlower = np.zeros((self.n,1))
        self.xupper = np.full((self.n,1),float('inf'))
        self.xlower[ix_['X1']] = 0.2
        self.xupper[ix_['X1']] = 4.0
        self.xlower[ix_['X2']] = 0.1
        self.xupper[ix_['X2']] = 1.6
        #%%%%%%%%%%%%%%%%%% START POINT %%%%%%%%%%%%%%%%%%
        self.x0 = np.full((self.n,1),float(1.0))
        #%%%%%%%%%%%%%%%%%%%% ELFTYPE %%%%%%%%%%%%%%%%%%%%%
        iet_  = {}
        elftv = []
        [it,iet_,_] = s2mpj_ii( 'eOE', iet_)
        elftv = loaset(elftv,it,0,'XX')
        elftv = loaset(elftv,it,1,'YY')
        [it,iet_,_] = s2mpj_ii( 'eCE1', iet_)
        elftv = loaset(elftv,it,0,'XX')
        elftv = loaset(elftv,it,1,'YY')
        [it,iet_,_] = s2mpj_ii( 'eCE2', iet_)
        elftv = loaset(elftv,it,0,'XX')
        elftv = loaset(elftv,it,1,'YY')
        #%%%%%%%%%%%%%%%%%% ELEMENT USES %%%%%%%%%%%%%%%%%%
        ie_ = {}
        self.elftype = np.array([])
        ielftype     = np.array([])
        self.elvar   = []
        ename = 'OBEL'
        [ie,ie_,_] = s2mpj_ii(ename,ie_)
        self.elftype = arrset(self.elftype,ie,'eOE')
        ielftype = arrset(ielftype,ie,iet_["eOE"])
        vname = 'X1'
        [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,float(1.0))
        posev = np.where(elftv[ielftype[ie]]=='XX')[0]
        self.elvar = loaset(self.elvar,ie,posev[0],iv)
        vname = 'X2'
        [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,float(1.0))
        posev = np.where(elftv[ielftype[ie]]=='YY')[0]
        self.elvar = loaset(self.elvar,ie,posev[0],iv)
        ename = 'COEL1'
        [ie,ie_,_] = s2mpj_ii(ename,ie_)
        self.elftype = arrset(self.elftype,ie,'eCE1')
        ielftype = arrset(ielftype,ie,iet_["eCE1"])
        vname = 'X1'
        [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,float(1.0))
        posev = np.where(elftv[ielftype[ie]]=='XX')[0]
        self.elvar = loaset(self.elvar,ie,posev[0],iv)
        vname = 'X2'
        [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,float(1.0))
        posev = np.where(elftv[ielftype[ie]]=='YY')[0]
        self.elvar = loaset(self.elvar,ie,posev[0],iv)
        ename = 'COEL2'
        [ie,ie_,_] = s2mpj_ii(ename,ie_)
        self.elftype = arrset(self.elftype,ie,'eCE2')
        ielftype = arrset(ielftype,ie,iet_["eCE2"])
        vname = 'X1'
        [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,float(1.0))
        posev = np.where(elftv[ielftype[ie]]=='XX')[0]
        self.elvar = loaset(self.elvar,ie,posev[0],iv)
        vname = 'X2'
        [iv,ix_] = s2mpj_nlx(self,vname,ix_,1,None,None,float(1.0))
        posev = np.where(elftv[ielftype[ie]]=='YY')[0]
        self.elvar = loaset(self.elvar,ie,posev[0],iv)
        #%%%%%%%%%%%%%%%%%%% GROUP USES %%%%%%%%%%%%%%%%%%%
        self.grelt   = []
        for ig in np.arange(0,ngrp):
            self.grelt.append(np.array([]))
        self.grftype = np.array([])
        self.grelw   = []
        nlc         = np.array([])
        ig = ig_['OBJ']
        posel = len(self.grelt[ig])
        self.grelt = loaset(self.grelt,ig,posel,ie_['OBEL'])
        nlc = np.union1d(nlc,np.array([ig]))
        self.grelw = loaset(self.grelw,ig,posel,1.)
        ig = ig_['CONS1']
        posel = len(self.grelt[ig])
        self.grelt = loaset(self.grelt,ig,posel,ie_['COEL1'])
        nlc = np.union1d(nlc,np.array([ig]))
        self.grelw = loaset(self.grelw,ig,posel,float(0.124))
        ig = ig_['CONS2']
        posel = len(self.grelt[ig])
        self.grelt = loaset(self.grelt,ig,posel,ie_['COEL2'])
        nlc = np.union1d(nlc,np.array([ig]))
        self.grelw = loaset(self.grelw,ig,posel,float(0.124))
        #%%%%%%%%%%%%%%%%%% OBJECT BOUNDS %%%%%%%%%%%%%%%%%
        self.objlower = 0.0
#    Solution
# LO SOLTN               1.5086379655
        #%%%%%%%% DEFAULT FOR MISSING SECTION(S) %%%%%%%%%%
        #%%%%%%%%%%%%% FORM clower AND cupper %%%%%%%%%%%%%
        self.clower = np.full((self.m,1),-float('Inf'))
        self.cupper = np.full((self.m,1),+float('Inf'))
        self.cupper[np.arange(self.nle)] = np.zeros((self.nle,1))
        #%%%% RETURN VALUES FROM THE __INIT__ METHOD %%%%%%
        self.lincons  = (
              np.where(np.isin(self.congrps,np.setdiff1d(self.congrps,nlc)))[0])
        self.pbclass   = "C-COOR2-MN-2-2"
        self.objderlvl = 2
        self.conderlvl = [2]

# **********************
#  SET UP THE FUNCTION *
#  AND RANGE ROUTINES  *
# **********************

    #%%%%%%%%%%%%%%% NONLINEAR ELEMENTS %%%%%%%%%%%%%%%

    @staticmethod
    def eOE(self, nargout,*args):

        import numpy as np
        EV_  = args[0]
        iel_ = args[1]
        A = 1.0+EV_[1,0]*EV_[1,0]
        RA = np.sqrt(A)
        f_   = EV_[0,0]*RA
        if nargout>1:
            try:
                dim = len(IV_)
            except:
                dim = len(EV_)
            g_ = np.zeros(dim)
            g_[0] = RA
            g_[1] = EV_[0,0]*EV_[1,0]/RA
            if nargout>2:
                H_ = np.zeros((2,2))
                H_[0,1] = EV_[1,0]/RA
                H_[1,0] = H_[0,1]
                H_[1,1] = EV_[0,0]/(A*RA)
        if nargout == 1:
            return f_
        elif nargout == 2:
            return f_,g_
        elif nargout == 3:
            return f_,g_,H_

    @staticmethod
    def eCE1(self, nargout,*args):

        import numpy as np
        EV_  = args[0]
        iel_ = args[1]
        A = 1.0+EV_[1,0]*EV_[1,0]
        RA = np.sqrt(A)
        B = 8.0/EV_[0,0]
        DB = -8.0/EV_[0,0]**2
        D2B = 16.0/EV_[0,0]**3
        C = 1.0/(EV_[0,0]*EV_[1,0])
        DCDX = -1.0/(EV_[0,0]**2*EV_[1,0])
        DCDY = -1.0/(EV_[1,0]**2*EV_[0,0])
        D2CDXX = 2.0/(EV_[0,0]**3*EV_[1,0])
        D2CDXY = 1.0/(EV_[0,0]*EV_[1,0])**2
        D2CDYY = 2.0/(EV_[0,0]*EV_[1,0]**3)
        BC = B+C
        f_   = RA*BC
        if nargout>1:
            try:
                dim = len(IV_)
            except:
                dim = len(EV_)
            g_ = np.zeros(dim)
            g_[0] = RA*(DB+DCDX)
            g_[1] = EV_[1,0]*BC/RA+RA*DCDY
            if nargout>2:
                H_ = np.zeros((2,2))
                H_[0,0] = RA*(D2B+D2CDXX)
                H_[0,1] = RA*D2CDXY+EV_[1,0]*(DB+DCDX)/RA
                H_[1,0] = H_[0,1]
                H_[1,1] = (BC+2.0*EV_[1,0]*DCDY-EV_[1,0]*EV_[1,0]*BC/A)/RA+RA*D2CDYY
        if nargout == 1:
            return f_
        elif nargout == 2:
            return f_,g_
        elif nargout == 3:
            return f_,g_,H_

    @staticmethod
    def eCE2(self, nargout,*args):

        import numpy as np
        EV_  = args[0]
        iel_ = args[1]
        A = 1.0+EV_[1,0]*EV_[1,0]
        RA = np.sqrt(A)
        B = 8.0/EV_[0,0]
        DB = -8.0/EV_[0,0]**2
        D2B = 16.0/EV_[0,0]**3
        C = 1.0/(EV_[0,0]*EV_[1,0])
        DCDX = -1.0/(EV_[0,0]**2*EV_[1,0])
        DCDY = -1.0/(EV_[1,0]**2*EV_[0,0])
        D2CDXX = 2.0/(EV_[0,0]**3*EV_[1,0])
        D2CDXY = 1.0/(EV_[0,0]*EV_[1,0])**2
        D2CDYY = 2.0/(EV_[0,0]*EV_[1,0]**3)
        BC = B-C
        f_   = RA*BC
        if nargout>1:
            try:
                dim = len(IV_)
            except:
                dim = len(EV_)
            g_ = np.zeros(dim)
            g_[0] = RA*(DB-DCDX)
            g_[1] = EV_[1,0]*BC/RA-RA*DCDY
            if nargout>2:
                H_ = np.zeros((2,2))
                H_[0,0] = RA*(D2B-D2CDXX)
                H_[0,1] = -RA*D2CDXY+EV_[1,0]*(DB-DCDX)/RA
                H_[1,0] = H_[0,1]
                H_[1,1] = (BC-2.0*EV_[1,0]*DCDY-EV_[1,0]*EV_[1,0]*BC/A)/RA-RA*D2CDYY
        if nargout == 1:
            return f_
        elif nargout == 2:
            return f_,g_
        elif nargout == 3:
            return f_,g_,H_

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

File: julia_problems/TWOBARS.jl
function TWOBARS(action::String,args::Union{PBM,Int,Float64,Vector{Int},Vector{Float64}}...)
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# 
# 
#    Problem:
#    ********
# 
#    Structureal analysis of the simplest two bar scheme.  The structure has
#    the following simple symmetric shape
# 
#                                 *
#                                / \
#                               /   \
#                              /     \
#                            """     """
# 
#    and a force is applied at the top node.  The unknown are the distance
#    of the left and right feet wrt to the projection of the top node and the
#    weight of the bars.
# 
#    Source:
#    an example in a talk by W.K. Zhang and C. Fleury, LLN, 1994.
# 
#    SIF input: Ph. Toint, November 1994
# 
#    classification = "C-OOR2-MN-2-2"
# 
# 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#   Translated to Julia by S2MPJ version 7 X 2024
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    name = "TWOBARS"

    if action == "setup"
        pb           = PB(name)
        pbm          = PBM(name)
        nargin       = length(args)
        pbm.call     = getfield( Main, Symbol( name ) )

        #%%%%%%%%%%%%%%%%%%%  PREAMBLE %%%%%%%%%%%%%%%%%%%%
        v_  = Dict{String,Float64}();
        ix_ = Dict{String,Int}();
        ig_ = Dict{String,Int}();
        #%%%%%%%%%%%%%%%%%%%  VARIABLES %%%%%%%%%%%%%%%%%%%%
        pb.xscale = Float64[]
        intvars = Int64[]
        binvars = Int64[]
        iv,ix_,_ = s2mpj_ii("X1",ix_)
        arrset(pb.xnames,iv,"X1")
        iv,ix_,_ = s2mpj_ii("X2",ix_)
        arrset(pb.xnames,iv,"X2")
        #%%%%%%%%%%%%%%%%%%  DATA GROUPS %%%%%%%%%%%%%%%%%%%
        gtype    = String[]
        ig,ig_,_ = s2mpj_ii("OBJ",ig_)
        arrset(gtype,ig,"<>")
        ig,ig_,_ = s2mpj_ii("CONS1",ig_)
        arrset(gtype,ig,"<=")
        arrset(pb.cnames,ig,"CONS1")
        ig,ig_,_ = s2mpj_ii("CONS2",ig_)
        arrset(gtype,ig,"<=")
        arrset(pb.cnames,ig,"CONS2")
        #%%%%%%%%%%%%%% GLOBAL DIMENSIONS %%%%%%%%%%%%%%%%%
        pb.n   = length(ix_)
        ngrp   = length(ig_)
        legrps = findall(x->x=="<=",gtype)
        eqgrps = findall(x->x=="==",gtype)
        gegrps = findall(x->x==">=",gtype)
        pb.nle = length(legrps)
        pb.neq = length(eqgrps)
        pb.nge = length(gegrps)
        pb.m   = pb.nle+pb.neq+pb.nge
        pbm.congrps = [[legrps;eqgrps];gegrps]
        pb.nob = ngrp-pb.m
        pbm.objgrps = findall(x->x=="<>",gtype)
        #%%%%%%%%%%%%%%%%%% CONSTANTS %%%%%%%%%%%%%%%%%%%%%
        pbm.gconst = zeros(Float64,ngrp)
        pbm.gconst[ig_["CONS1"]] = Float64(1.0)
        pbm.gconst[ig_["CONS2"]] = Float64(1.0)
        #%%%%%%%%%%%%%%%%%%%%  BOUNDS %%%%%%%%%%%%%%%%%%%%%
        pb.xlower = zeros(Float64,pb.n)
        pb.xupper =    fill(Inf,pb.n)
        pb.xlower[ix_["X1"]] = 0.2
        pb.xupper[ix_["X1"]] = 4.0
        pb.xlower[ix_["X2"]] = 0.1
        pb.xupper[ix_["X2"]] = 1.6
        #%%%%%%%%%%%%%%%%%% START POINT %%%%%%%%%%%%%%%%%%
        pb.x0 = fill(Float64(1.0),pb.n)
        #%%%%%%%%%%%%%%%%%%%% ELFTYPE %%%%%%%%%%%%%%%%%%%%%
        iet_  = Dict{String,Int}()
        elftv = Vector{Vector{String}}()
        it,iet_,_ = s2mpj_ii( "eOE", iet_)
        loaset(elftv,it,1,"XX")
        loaset(elftv,it,2,"YY")
        it,iet_,_ = s2mpj_ii( "eCE1", iet_)
        loaset(elftv,it,1,"XX")
        loaset(elftv,it,2,"YY")
        it,iet_,_ = s2mpj_ii( "eCE2", iet_)
        loaset(elftv,it,1,"XX")
        loaset(elftv,it,2,"YY")
        #%%%%%%%%%%%%%%%%%% ELEMENT USES %%%%%%%%%%%%%%%%%%
        ie_      = Dict{String,Int}()
        ielftype = Vector{Int64}()
        ename = "OBEL"
        ie,ie_,_  = s2mpj_ii(ename,ie_)
        arrset(pbm.elftype,ie,"eOE")
        arrset(ielftype,ie,iet_["eOE"])
        vname = "X1"
        iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,Float64(1.0))
        posev = findfirst(x->x=="XX",elftv[ielftype[ie]])
        loaset(pbm.elvar,ie,posev,iv)
        vname = "X2"
        iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,Float64(1.0))
        posev = findfirst(x->x=="YY",elftv[ielftype[ie]])
        loaset(pbm.elvar,ie,posev,iv)
        ename = "COEL1"
        ie,ie_,_  = s2mpj_ii(ename,ie_)
        arrset(pbm.elftype,ie,"eCE1")
        arrset(ielftype,ie,iet_["eCE1"])
        vname = "X1"
        iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,Float64(1.0))
        posev = findfirst(x->x=="XX",elftv[ielftype[ie]])
        loaset(pbm.elvar,ie,posev,iv)
        vname = "X2"
        iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,Float64(1.0))
        posev = findfirst(x->x=="YY",elftv[ielftype[ie]])
        loaset(pbm.elvar,ie,posev,iv)
        ename = "COEL2"
        ie,ie_,_  = s2mpj_ii(ename,ie_)
        arrset(pbm.elftype,ie,"eCE2")
        arrset(ielftype,ie,iet_["eCE2"])
        vname = "X1"
        iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,Float64(1.0))
        posev = findfirst(x->x=="XX",elftv[ielftype[ie]])
        loaset(pbm.elvar,ie,posev,iv)
        vname = "X2"
        iv,ix_,pb = s2mpj_nlx(vname,ix_,pb,1,nothing,nothing,Float64(1.0))
        posev = findfirst(x->x=="YY",elftv[ielftype[ie]])
        loaset(pbm.elvar,ie,posev,iv)
        #%%%%%%%%%%%%%%%%%%% GROUP USES %%%%%%%%%%%%%%%%%%%
        for ig in 1:ngrp
            arrset(pbm.grelt,ig,Int64[])
        end
        nlc = Int64[]
        ig = ig_["OBJ"]
        posel = length(pbm.grelt[ig])+1
        loaset(pbm.grelt,ig,posel,ie_["OBEL"])
        arrset(nlc,length(nlc)+1,ig)
        loaset(pbm.grelw,ig,posel,1.)
        ig = ig_["CONS1"]
        posel = length(pbm.grelt[ig])+1
        loaset(pbm.grelt,ig,posel,ie_["COEL1"])
        arrset(nlc,length(nlc)+1,ig)
        loaset(pbm.grelw,ig,posel,Float64(0.124))
        ig = ig_["CONS2"]
        posel = length(pbm.grelt[ig])+1
        loaset(pbm.grelt,ig,posel,ie_["COEL2"])
        arrset(nlc,length(nlc)+1,ig)
        loaset(pbm.grelw,ig,posel,Float64(0.124))
        #%%%%%%%%%%%%%%%%%% OBJECT BOUNDS %%%%%%%%%%%%%%%%%
        pb.objlower = 0.0
#    Solution
# LO SOLTN               1.5086379655
        #%%%%%%%% DEFAULT FOR MISSING SECTION(S) %%%%%%%%%%
        #%%%%%%%%%%%%% FORM clower AND cupper %%%%%%%%%%%%%
        pb.clower = -1*fill(Inf,pb.m)
        pb.cupper =    fill(Inf,pb.m)
        pb.cupper[1:pb.nle] = zeros(Float64,pb.nle)
        pbm.A = spzeros(Float64,0,0)
        pbm.H = spzeros(Float64,0,0)
        #%%%%% RETURN VALUES FROM THE SETUP ACTION %%%%%%%%
        pb.lincons = findall(x-> x in setdiff( pbm.congrps,nlc),pbm.congrps)
        pb.pbclass = "C-OOR2-MN-2-2"
        pbm.objderlvl = 2
        pb.objderlvl = pbm.objderlvl;
        pbm.conderlvl = [2]
        pb.conderlvl  = pbm.conderlvl;
        return pb, pbm

# **********************
#  SET UP THE FUNCTION *
#  AND RANGE ROUTINES  *
# **********************

    #%%%%%%%%%%%%%%% NONLINEAR ELEMENTS %%%%%%%%%%%%%%%

    elseif action == "eOE"

        EV_     = args[1]
        iel_    = args[2]
        nargout = args[3]
        pbm     = args[4]
        A = 1.0+EV_[2]*EV_[2]
        RA = sqrt(A)
        f_   = EV_[1]*RA
        if nargout>1
            dim = try length(IV_) catch; length(EV_) end
            g_  = zeros(Float64,dim)
            g_[1] = RA
            g_[2] = EV_[1]*EV_[2]/RA
            if nargout>2
                H_ = zeros(Float64,2,2)
                H_[1,2] = EV_[2]/RA
                H_[2,1] = H_[1,2]
                H_[2,2] = EV_[1]/(A*RA)
            end
        end
        if nargout == 1
            return f_
        elseif nargout == 2
            return f_,g_
        elseif nargout == 3
            return f_,g_,H_
        end

    elseif action == "eCE1"

        EV_     = args[1]
        iel_    = args[2]
        nargout = args[3]
        pbm     = args[4]
        A = 1.0+EV_[2]*EV_[2]
        RA = sqrt(A)
        B = 8.0/EV_[1]
        DB = -8.0/EV_[1]^2
        D2B = 16.0/EV_[1]^3
        C = 1.0/(EV_[1]*EV_[2])
        DCDX = -1.0/(EV_[1]^2*EV_[2])
        DCDY = -1.0/(EV_[2]^2*EV_[1])
        D2CDXX = 2.0/(EV_[1]^3*EV_[2])
        D2CDXY = 1.0/(EV_[1]*EV_[2])^2
        D2CDYY = 2.0/(EV_[1]*EV_[2]^3)
        BC = B+C
        f_   = RA*BC
        if nargout>1
            dim = try length(IV_) catch; length(EV_) end
            g_  = zeros(Float64,dim)
            g_[1] = RA*(DB+DCDX)
            g_[2] = EV_[2]*BC/RA+RA*DCDY
            if nargout>2
                H_ = zeros(Float64,2,2)
                H_[1,1] = RA*(D2B+D2CDXX)
                H_[1,2] = RA*D2CDXY+EV_[2]*(DB+DCDX)/RA
                H_[2,1] = H_[1,2]
                H_[2,2] = (BC+2.0*EV_[2]*DCDY-EV_[2]*EV_[2]*BC/A)/RA+RA*D2CDYY
            end
        end
        if nargout == 1
            return f_
        elseif nargout == 2
            return f_,g_
        elseif nargout == 3
            return f_,g_,H_
        end

    elseif action == "eCE2"

        EV_     = args[1]
        iel_    = args[2]
        nargout = args[3]
        pbm     = args[4]
        A = 1.0+EV_[2]*EV_[2]
        RA = sqrt(A)
        B = 8.0/EV_[1]
        DB = -8.0/EV_[1]^2
        D2B = 16.0/EV_[1]^3
        C = 1.0/(EV_[1]*EV_[2])
        DCDX = -1.0/(EV_[1]^2*EV_[2])
        DCDY = -1.0/(EV_[2]^2*EV_[1])
        D2CDXX = 2.0/(EV_[1]^3*EV_[2])
        D2CDXY = 1.0/(EV_[1]*EV_[2])^2
        D2CDYY = 2.0/(EV_[1]*EV_[2]^3)
        BC = B-C
        f_   = RA*BC
        if nargout>1
            dim = try length(IV_) catch; length(EV_) end
            g_  = zeros(Float64,dim)
            g_[1] = RA*(DB-DCDX)
            g_[2] = EV_[2]*BC/RA-RA*DCDY
            if nargout>2
                H_ = zeros(Float64,2,2)
                H_[1,1] = RA*(D2B-D2CDXX)
                H_[1,2] = -RA*D2CDXY+EV_[2]*(DB-DCDX)/RA
                H_[2,1] = H_[1,2]
                H_[2,2] = (BC-2.0*EV_[2]*DCDY-EV_[2]*EV_[2]*BC/A)/RA-RA*D2CDYY
            end
        end
        if nargout == 1
            return f_
        elseif nargout == 2
            return f_,g_
        elseif nargout == 3
            return f_,g_,H_
        end

    #%%%%%%%%%%%%%%% THE MAIN ACTIONS %%%%%%%%%%%%%%%

    elseif action in  ["fx","fgx","fgHx","cx","cJx","cJHx","cIx","cIJx","cIJHx","cIJxv","fHxv",
                       "cJxv","cJtxv","cIJtxv","Lxy","Lgxy","LgHxy","LIxy","LIgxy","LIgHxy",
                       "LHxyv","LIHxyv"]

        pbm = args[1]
        if pbm.name == name
            pbm.has_globs = [0,0]
            return s2mpj_eval(action,args...)
        else
            println("ERROR: please run "*name*" with action = setup")
            return ntuple(i->undef,args[end])
        end

    else
        println("ERROR: action "*action*" unavailable for problem "*name*".jl")
        return ntuple(i->undef,args[end])
    end

end

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

